// firebase-tools/lib/dataconnect/fileUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSdkYaml = exports.directoryHasPackageJson = exports.getPlatformFromFolder = exports.pickService = exports.readGQLFiles = exports.readConnectorYaml = exports.readDataConnectYaml = exports.readFirebaseJson = void 0;
const fs = require("fs-extra");
const path = require("path");
const error_1 = require("../error");
const types_1 = require("./types");
const utils_1 = require("../utils");
const load_1 = require("./load");
function readFirebaseJson(config) {
    if (!config.has("dataconnect")) {
        return [];
    }
    const validator = (cfg) => {
        if (!cfg["source"]) {
            throw new error_1.FirebaseError("Invalid firebase.json: DataConnect requires `source`");
        }
        return {
            source: cfg["source"],
        };
    };
    const configs = config.get("dataconnect");
    if (typeof configs === "object" && !Array.isArray(configs)) {
        return [validator(configs)];
    }
    else if (Array.isArray(configs)) {
        return configs.map(validator);
    }
    else {
        throw new error_1.FirebaseError("Invalid firebase.json: dataconnect should be of the form { source: string }");
    }
}
exports.readFirebaseJson = readFirebaseJson;
async function readDataConnectYaml(sourceDirectory) {
    const file = await (0, utils_1.readFileFromDirectory)(sourceDirectory, "dataconnect.yaml");
    const dataconnectYaml = await (0, utils_1.wrappedSafeLoad)(file.source);
    return validateDataConnectYaml(dataconnectYaml);
}
exports.readDataConnectYaml = readDataConnectYaml;
function validateDataConnectYaml(unvalidated) {
    if (!unvalidated["location"]) {
        throw new error_1.FirebaseError("Missing required field 'location' in dataconnect.yaml");
    }
    return unvalidated;
}
async function readConnectorYaml(sourceDirectory) {
    const file = await (0, utils_1.readFileFromDirectory)(sourceDirectory, "connector.yaml");
    const connectorYaml = await (0, utils_1.wrappedSafeLoad)(file.source);
    return validateConnectorYaml(connectorYaml);
}
exports.readConnectorYaml = readConnectorYaml;
function validateConnectorYaml(unvalidated) {
    return unvalidated;
}
async function readGQLFiles(sourceDir) {
    const files = await fs.readdir(sourceDir);
    return files
        .filter((f) => f.endsWith(".gql") || f.endsWith(".graphql"))
        .map((f) => toFile(sourceDir, f));
}
exports.readGQLFiles = readGQLFiles;
function toFile(sourceDir, relPath) {
    const fullPath = path.join(sourceDir, relPath);
    if (!fs.existsSync(fullPath)) {
        throw new error_1.FirebaseError(`file ${fullPath} not found`);
    }
    const content = fs.readFileSync(fullPath).toString();
    return {
        path: relPath,
        content,
    };
}
async function pickService(projectId, config, serviceId) {
    const serviceCfgs = readFirebaseJson(config);
    let serviceInfo;
    if (serviceCfgs.length === 0) {
        throw new error_1.FirebaseError("No Data Connect services found in firebase.json.");
    }
    else if (serviceCfgs.length === 1) {
        serviceInfo = await (0, load_1.load)(projectId, config, serviceCfgs[0].source);
    }
    else {
        if (!serviceId) {
            throw new error_1.FirebaseError("Multiple Data Connect services found in firebase.json. Please specify a service ID to use.");
        }
        const infos = await Promise.all(serviceCfgs.map((c) => (0, load_1.load)(projectId, config, c.source)));
        const maybe = infos.find((i) => i.dataConnectYaml.serviceId === serviceId);
        if (!maybe) {
            throw new error_1.FirebaseError(`No service named ${serviceId} declared in firebase.json.`);
        }
        serviceInfo = maybe;
    }
    return serviceInfo;
}
exports.pickService = pickService;
const WEB_INDICATORS = ["package.json", "package-lock.json", "node_modules"];
const IOS_INDICATORS = ["info.plist", "podfile", "package.swift"];
const ANDROID_INDICATORS = ["androidmanifest.xml", "build.gradle", "build.gradle.kts"];
const IOS_POSTFIX_INDICATORS = [".xcworkspace", ".xcodeproj"];
async function getPlatformFromFolder(dirPath) {
    const fileNames = await fs.readdir(dirPath);
    let hasWeb = false;
    let hasAndroid = false;
    let hasIOS = false;
    for (const fileName of fileNames) {
        const cleanedFileName = fileName.toLowerCase();
        hasWeb || (hasWeb = WEB_INDICATORS.some((indicator) => indicator === cleanedFileName));
        hasAndroid || (hasAndroid = ANDROID_INDICATORS.some((indicator) => indicator === cleanedFileName));
        hasIOS || (hasIOS = IOS_INDICATORS.some((indicator) => indicator === cleanedFileName) ||
            IOS_POSTFIX_INDICATORS.some((indicator) => cleanedFileName.endsWith(indicator)));
    }
    if (hasWeb && !hasAndroid && !hasIOS) {
        return types_1.Platform.WEB;
    }
    else if (hasAndroid && !hasWeb && !hasIOS) {
        return types_1.Platform.ANDROID;
    }
    else if (hasIOS && !hasWeb && !hasAndroid) {
        return types_1.Platform.IOS;
    }
    return types_1.Platform.UNDETERMINED;
}
exports.getPlatformFromFolder = getPlatformFromFolder;
async function directoryHasPackageJson(dirPath) {
    const fileNames = await fs.readdir(dirPath);
    return fileNames.some((f) => f.toLowerCase() === "package.json");
}
exports.directoryHasPackageJson = directoryHasPackageJson;
function generateSdkYaml(platform, connectorYaml, connectorYamlFolder, appFolder) {
    const relPath = path.relative(connectorYamlFolder, appFolder);
    const outputDir = path.join(relPath, "dataconnect-generated");
    if (!connectorYaml.generate) {
        connectorYaml.generate = {};
    }
    if (platform === types_1.Platform.WEB) {
        connectorYaml.generate.javascriptSdk = {
            outputDir,
            package: `@firebasegen/${connectorYaml.connectorId}`,
            packageJsonDir: appFolder,
        };
    }
    if (platform === types_1.Platform.IOS) {
        connectorYaml.generate.swiftSdk = {
            outputDir,
            package: connectorYaml.connectorId,
        };
    }
    if (platform === types_1.Platform.ANDROID) {
        connectorYaml.generate.kotlinSdk = {
            outputDir,
            package: `connectors.${connectorYaml.connectorId}`,
        };
    }
    return connectorYaml;
}
exports.generateSdkYaml = generateSdkYaml;
