// firebase-tools/lib/emulator/apphosting/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppHostingEmulator = void 0;
const emulatorLogger_1 = require("../emulatorLogger");
const types_1 = require("../types");
const serve_1 = require("./serve");
class AppHostingEmulator {
    constructor(args) {
        this.args = args;
        this.logger = emulatorLogger_1.EmulatorLogger.forEmulator(types_1.Emulators.APPHOSTING);
    }
    async start() {
        this.args.options.host = this.args.host;
        this.args.options.port = this.args.port;
        this.logger.logLabeled("INFO", types_1.Emulators.APPHOSTING, "starting apphosting emulator");
        const { port } = await (0, serve_1.start)(this.args.options);
        this.logger.logLabeled("INFO", types_1.Emulators.APPHOSTING, `serving on port ${port}`);
    }
    connect() {
        this.logger.logLabeled("INFO", types_1.Emulators.APPHOSTING, "connecting apphosting emulator");
        return Promise.resolve();
    }
    stop() {
        this.logger.logLabeled("INFO", types_1.Emulators.APPHOSTING, "stopping apphosting emulator");
        return Promise.resolve();
    }
    getInfo() {
        return {
            name: types_1.Emulators.APPHOSTING,
            host: this.args.host,
            port: this.args.port,
        };
    }
    getName() {
        return types_1.Emulators.APPHOSTING;
    }
}
exports.AppHostingEmulator = AppHostingEmulator;
